services/headless-lms/chatbot/src/chatbot_tools/course_progress.rs

```diff
@@ -0,0 +1,320 @@
+        res += &format!("On this {course_or_module}, there are a total of ");
+
+        if let Some(a) = total_exercises {
+            res += &format!("{a} exercises");
+        }
+        if let Some(b) = score_maximum {
+            if total_exercises.is_some() {
+                res += " and ";
+            }
+            res += &format!("{b} exercise points");
+        }
+        res += ". ";
+    }
+
+    if attempted_exercises_required.is_some() || score_required.is_some() {
+        res += &format!("To pass this {course_or_module}, it's required to ");
+
+        if let Some(a) = attempted_exercises_required {
+            res += &format!("attempt {a} exercises");
+        }
+        if let Some(b) = score_required {
+            if attempted_exercises_required.is_some() {
+                res += " and ";
+            }
+            res += &format!("gain {b} exercise points");
+        }
+        res += ". ";
+    }
+
+    if let Some(b) = attempted_exercises {
+        res += &format!("The user has attempted {b} exercises. ");
+    } else {
+        res += "The user has not attempted any exercises. ";
+    }
+    let attempted_exercises_n = attempted_exercises.unwrap_or(0);
+
+    if let Some(c) = attempted_exercises_required {
+        let left = c - attempted_exercises_n;
+        if left <= 0 {
+            res += &format!(
+                "They have attempted enough exercises to pass this {course_or_module} if they have also received enough points. "
+            );
+        } else {
+            res += &format!(
+                "To pass this {course_or_module}, they need to attempt {left} more exercises. "
+            );
+        }
+    }
+
+    res += &format!("The user has gained {:.1} points. ", score_given);
+
+    if let Some(e) = score_required {
+        let left = e as f32 - score_given;
+        if left <= 0 as f32 {
+            res += &format!("The user has gained enough points to pass this {course_or_module}. ")
+        } else {
+            res += &format!(
+                "To pass this {course_or_module}, the user needs to gain {:.1} more points. ",
+                left
+            )
+        }
+    }
+    res + "\n"
+}
+
+#[cfg(test)]
+mod tests {
+    use uuid::Uuid;
+
+    use super::*;
+
+    impl CourseProgressTool {
+        fn new_mock(course_name: String, progress: Vec<UserCourseProgress>) -> Self {
+            CourseProgressTool {
+                state: CourseProgressState {
+                    course_name,
+                    progress,
+                },
+                arguments: CourseProgressArguments {},
+            }
+        }
+    }
+
+    #[test]
+    fn test_course_progress_output_only_base_module() {
+        let progress = vec![UserCourseProgress {
+            course_module_id: Uuid::nil(),
+            course_module_name: "Example base module".to_string(),
+            course_module_order_number: 1,
+            score_given: 3.3,
```

Comment:
Try returning more than 1 decimal and then test that the score_given has been rounded down to one decimal.